src/sqlancer/ASTBasedReducer.java

```diff
@@ -0,0 +1,470 @@
+package sqlancer;
+
+import net.sf.jsqlparser.expression.*;
+import net.sf.jsqlparser.expression.operators.relational.*;
+import net.sf.jsqlparser.parser.CCJSqlParserUtil;
+import net.sf.jsqlparser.statement.Statement;
+import net.sf.jsqlparser.statement.StatementVisitorAdapter;
+import net.sf.jsqlparser.statement.select.*;
+import sqlancer.common.query.Query;
+import sqlancer.common.query.SQLQueryAdapter;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.*;
+import java.util.function.BiConsumer;
+
+class ExpressionTransformer {
+
+    private static List<Expression> transformExpr(BinaryExpression expr) {
+        List<Expression> candidates = new ArrayList<>();
+        Expression lhs = (expr).getLeftExpression();
+        Expression rhs = (expr).getRightExpression();
+        candidates.add(lhs);
+        candidates.add(rhs);
+        return candidates;
+    }
+
+    private static List<Expression> transformExpr(Between expr) {
+        Expression lhs = expr.getBetweenExpressionStart();
+        Expression rhs = expr.getBetweenExpressionEnd();
+        return List.of(lhs, rhs);
+    }
+
+    public static List<Expression> candidateExpressions(Expression expr) {
+
+        // Queue<Expression> candidateExprQue = new PriorityQueue<>(Comparator.comparingInt(o ->
+        // o.toString().length()));
+        if (expr instanceof Parenthesis) {
+            // try removing a pair of brackets.
+            Parenthesis paren = (Parenthesis) expr;
+            return List.of(paren.getExpression());
+        } else if (expr instanceof BinaryExpression) {
+            return transformExpr((BinaryExpression) expr);
+        } else if (expr instanceof Between) {
+            return transformExpr((Between) expr);
+        } else if (expr instanceof LongValue) {
+            LongValue longValue = (LongValue) expr;
+            if (longValue.getValue() >= 1000) {
+                return List.of(new NullValue(), new LongValue(10), new LongValue(0), new LongValue(1));
+            }
+            return new ArrayList<>();
+        } else if (expr instanceof DoubleValue) {
+            DoubleValue doubleValue = (DoubleValue) expr;
+            double literal = doubleValue.getValue();
+            if (String.valueOf(literal).length() <= 4) {
+                return new ArrayList<>();
+            }
+            double roundedValue = Math.round(literal * 10.0) / 10.0;
+            return List.of(new NullValue(), new DoubleValue(String.valueOf(roundedValue)));
+        } else if (expr instanceof CaseExpression) {
+            CaseExpression caseExpression = (CaseExpression) expr;
+            return List.of(caseExpression.getSwitchExpression(), caseExpression.getElseExpression());
+        } else
```

Comment:
Do we need to shorten other constant value like String?

Reply:
Yes, String value could be reduction. I would add it in the later PR.